src/lib/README.ctable.md

```diff
@@ -0,0 +1,198 @@
+collisions are possible as well of course; in that case we continue
+scanning forward.
+
+The distance travelled while scanning for the matching hash is known as
+the /displacement/.  The table measures its maximum displacement, for a
+number of purposes, but you might be interested to know that a maximum
+displacement for a table with 2 million entries and a 40% load factor is
+around 8 or 9.  Smaller tables will have smaller maximum displacements.
+
+The ctable has two lookup interfaces.  One will perform the lookup as
+described above, scanning through the hash table in place.  The other
+will fetch all entries within the maximum displacement into a buffer,
+then do a branchless binary search over that buffer.  This second
+streaming lookup can also fetch entries for multiple keys in one go.
+This can amortize the cost of a round-trip to RAM, in the case where you
+expect to miss cache for every lookup.
+
+To create a ctable, first create a parameters table specifying the key
+and value types, along with any other options.  Then call `ctable.new`
+on those parameters.  For example:
+
+```lua
+local ctable = require('lib.ctable')
+local ffi = require('ffi')
+local params = {
+   key_type = ffi.typeof('uint32_t'),
+   value_type = ffi.typeof('int32_t[6]'),
+   hash_fn = ctable.hash_i32,
+   max_occupancy_rate = 0.4,
+   initial_size = math.ceil(occupancy / 0.4)
+}
+local ctab = ctable.new(params)
+```
+
+— Function **ctable.new** *params*
+
+Create a new ctable.  *params* is a table of key/value pairs.  The
+following keys are required:
+
+ * `key_type`: An FFI type for keys in this table.
+ * `value_type`: An FFI type for values in this table.  (In the future,
+   `value_type` will be optional; a nil `value_type` will create a
+   set).
+ * `hash_fn`: A function that takes a key and returns a hash value.
+
+Hash values are unsigned 32-bit integers in the range `[0,
+0xFFFFFFFF)`.  That is to say, `0xFFFFFFFF` is the only unsigned 32-bit
+integer that is not a valid hash value.  The `hash_fn` must return a
+hash value in the correct range.
+
+Optional entries that may be present in the *params* table include:
+
+ * `initial_size`: The initial size of the hash table, including free
+   space.  Defaults to 8 slots.
+ * `max_occupancy_rate`: The maximum ratio of `occupancy/size`, where
+   `occupancy` denotes the number of entries in the table, and `size` is
+   the total table size including free entries.  Trying to add an entry
+   to a "full" table will cause the table to grow in size by a factor of
+   2.  Defaults to 0.9, for a 90% maximum occupancy ratio.
+ * `min_occupancy_rate`: Minimum ratio of `occupancy/size`.  Removing an
+   entry from an "empty" table will shrink the table.
+
+#### Methods
+
+Users interact with a ctable through methods.  In these method
+descriptions, the object on the left-hand-side of the method invocation
+should be a ctable.
+
+— Method **:resize** *size*
```

Comment:
> Prepend method definitions with the class name. E.g. ctable:new instead of :new

So, `new` is a function, but there are many methods like `resize`.  I was a bit worried though that there would be confusion between the module name and the class name, and for that reason left off the LHS on these definitions.  In fact partly for that reason we don't mention the name "class" in these docs, as it doesn't really exist.  What's the right solution here?  Is this acceptable as I've described it above?